lib/puppet/resource_api/property.rb

```diff
@@ -69,10 +83,25 @@ def rs_value
   # method overloaded only for the :ensure property, add option to check if the
   # rs_value matches is. Only if the class is child of
   # Puppet::ResourceApi::Property.
-  def def_insync?
+  def def_ensure_insync?
     define_singleton_method(:insync?) { |is| rs_value.to_s == is.to_s }
   end
 
+  def def_custom_insync?
+    define_singleton_method(:insync?) do |is|
+      provider    = @referrable_type.my_provider
+      context     = @referrable_type.context
+      should_hash = @resource.rsapi_canonicalized_target_state
+      is_hash     = @resource.rsapi_current_state
+      title       = @resource.rsapi_title
+
+      provider_insync_result = provider.insync?(context, title, @attribute_name, is_hash, should_hash) if provider.respond_to?(:insync?)
```

Comment:
If the custom resource has declared 'custom_insync' feature, it MUST provide the `insync?` method, no need to check (although a friendly error message might be useful).